net: ieee802154: Sort out fragmentation tests

rlubos opened the pull request:
Fragmentation tests in `tests/net/ieee802154` failed due to asserts on LL addresses and a memory corruption issue. This appears to be happening since the recent 6lowpan rework.

@jukkar It seems that these tests are not automatically executed during the sanity-check, that's why it went so long unnoticed. A short investigation showed up that the ieee802514 tests have the following dependency: `depends_on: ieee802154`, yet no emulated qemu board sets `ieee802154` in their capabilities. IMO we could fix it in this PR as well. Shall we add the capability to any of the boards or remove the dependency? The latter should be no harm, as the tests seem to provide either dummy network interface or dummy ieee802154 driver.

Fixes #19761.

rlubos committed:
net: 6lo: Fix memory corruption during uncompression

A memory corruption could happen in `uncompress_IPHC_header` function,
when data was moved to make place in the net buffer for the uncompressed
IPv6/UDP header.

The size of data being moved should only contain the original data size,
not incremented by the amount of space needed to expand the header,
which was already added to the net buffer size. In result, the `memmove`
operation could exceed the allocated net buffer and cause memory
corruption.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: ieee802154: Set LL addresses on a packet

The reworked 6lowpan implementation asserts when no LL source or
destination address is set on a packet. This caused the fragmentation
tests to fail before they completed.

Fix the issue, by setting a dummy address on a packet during a setup.

Fixes #19761

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

tbursztyka commented on tests/net/ieee802154/fragment/src/main.c:
```diff
@@ -287,6 +287,17 @@ static struct net_pkt *create_pkt(struct net_fragment_data *data)
 	}
 
+	/* Setup link layer addresses. */
+	static u16_t dummy_short_addr;
```
no found of this, but up to you to change. Seems to be accepted in zephyr...

rlubos replied:
> no found of this, but up to you to change

I didn't catch this up, could you elaborate?

tbursztyka replied:
well it seems to be accepted to declare variable in a middle of a code block.

rlubos replied:
Ah right, I can move it upwards, no prob

tbursztyka reviewed (approved):
wasn't an issue opened about it ? (or maybe was is just in a  slack talk?)

rlubos commented:
> wasn't an issue opened about it ? (or maybe was is just in a slack talk?)

Actually I'm not aware of either, I just found the problem out when running on hardware tests on nRF52840.

rlubos commented:
Oh, this actually could be the issue: https://github.com/zephyrproject-rtos/zephyr/issues/19761. I'll mention it in the original post.

jukkar reviewed (approved):
LGTM. Perhaps you could add "Fixes #19761" to commit message too.

jukkar commented:
> IMO we could fix it in this PR as well. Shall we add the capability to any of the boards or remove the dependency? The latter should be no harm, as the tests seem to provide either dummy network interface or dummy ieee802154 driver.

Agreed, we should make the tests run by sanitycheck automatically so I suppose we would need to add the capability to some board.

rlubos commented:
According to https://github.com/zephyrproject-rtos/zephyr/commit/de8d755c5403d7ac43bfd8c4c5860abb33c1a3a4, it seems that `qemu_x86` is favoured for the networking tests, I'll extend its capabilities then.

alexanderwachter reviewed (approved):
Thanks for the 6lo fix :-)

rlubos committed:
boards: qemu_x86: Add ieee802154 capability

Extend `qemu_x86` board configuration with `ieee802154` capability, to
allow it to execute `ieee802154` tests.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>
